### Sudoku Notation and Preliminary Notions
This part takes cares of the representation for the sudoku

In [2]:
def cross(A, B):
    "Cross product of elements in A and elements in B."
    return [a+b for a in A for b in B]

digits   = '123456789'
rows     = 'ABCDEFGHI'
cols     = digits
squares  = cross(rows, cols)
unitlist = ([cross(rows, c) for c in cols] +
            [cross(r, cols) for r in rows] +
            [cross(rs, cs) for rs in ('ABC','DEF','GHI') for cs in ('123','456','789')])
units = dict((s, [u for u in unitlist if s in u]) 
             for s in squares)
peers = dict((s, set(sum(units[s],[]))-set([s]))
             for s in squares)

def test():
    "A set of unit tests."
    assert len(squares) == 81
    assert len(unitlist) == 27
    assert all(len(units[s]) == 3 for s in squares)
    assert all(len(peers[s]) == 20 for s in squares)
    assert units['C2'] == [['A2', 'B2', 'C2', 'D2', 'E2', 'F2', 'G2', 'H2', 'I2'],
                           ['C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9'],
                           ['A1', 'A2', 'A3', 'B1', 'B2', 'B3', 'C1', 'C2', 'C3']]
    assert peers['C2'] == set(['A2', 'B2', 'D2', 'E2', 'F2', 'G2', 'H2', 'I2',
                               'C1', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9',
                               'A1', 'A3', 'B1', 'B3'])
    print 'All tests pass.'
    
test()

All tests pass.


### Debugging
To understand what those few lines of codes were trying to do.

In [14]:
print ([cross(rows, c) for c in cols]) 
print ()
print ([cross(r, cols) for r in rows])
print ()
print ([cross(rs, cs) for rs in ('ABC','DEF','GHI') for cs in ('123','456','789')])

[['A1', 'B1', 'C1', 'D1', 'E1', 'F1', 'G1', 'H1', 'I1'], ['A2', 'B2', 'C2', 'D2', 'E2', 'F2', 'G2', 'H2', 'I2'], ['A3', 'B3', 'C3', 'D3', 'E3', 'F3', 'G3', 'H3', 'I3'], ['A4', 'B4', 'C4', 'D4', 'E4', 'F4', 'G4', 'H4', 'I4'], ['A5', 'B5', 'C5', 'D5', 'E5', 'F5', 'G5', 'H5', 'I5'], ['A6', 'B6', 'C6', 'D6', 'E6', 'F6', 'G6', 'H6', 'I6'], ['A7', 'B7', 'C7', 'D7', 'E7', 'F7', 'G7', 'H7', 'I7'], ['A8', 'B8', 'C8', 'D8', 'E8', 'F8', 'G8', 'H8', 'I8'], ['A9', 'B9', 'C9', 'D9', 'E9', 'F9', 'G9', 'H9', 'I9']]
()
[['A1', 'A2', 'A3', 'A4', 'A5', 'A6', 'A7', 'A8', 'A9'], ['B1', 'B2', 'B3', 'B4', 'B5', 'B6', 'B7', 'B8', 'B9'], ['C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9'], ['D1', 'D2', 'D3', 'D4', 'D5', 'D6', 'D7', 'D8', 'D9'], ['E1', 'E2', 'E3', 'E4', 'E5', 'E6', 'E7', 'E8', 'E9'], ['F1', 'F2', 'F3', 'F4', 'F5', 'F6', 'F7', 'F8', 'F9'], ['G1', 'G2', 'G3', 'G4', 'G5', 'G6', 'G7', 'G8', 'G9'], ['H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'H7', 'H8', 'H9'], ['I1', 'I2', 'I3', 'I4', 'I5', 'I6', 'I7', 

In [34]:
for s in squares:
    print (s)
    print (units[s])
    print (set(sum(units[s],[]))- set([s]))
    break

A1
[['A1', 'B1', 'C1', 'D1', 'E1', 'F1', 'G1', 'H1', 'I1'], ['A1', 'A2', 'A3', 'A4', 'A5', 'A6', 'A7', 'A8', 'A9'], ['A1', 'A2', 'A3', 'B1', 'B2', 'B3', 'C1', 'C2', 'C3']]
set(['F1', 'C2', 'A7', 'G1', 'I1', 'H1', 'A9', 'A3', 'A2', 'A5', 'B1', 'B2', 'B3', 'C3', 'A8', 'C1', 'D1', 'E1', 'A6', 'A4'])


In [3]:
print (squares)

['A1', 'A2', 'A3', 'A4', 'A5', 'A6', 'A7', 'A8', 'A9', 'B1', 'B2', 'B3', 'B4', 'B5', 'B6', 'B7', 'B8', 'B9', 'C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9', 'D1', 'D2', 'D3', 'D4', 'D5', 'D6', 'D7', 'D8', 'D9', 'E1', 'E2', 'E3', 'E4', 'E5', 'E6', 'E7', 'E8', 'E9', 'F1', 'F2', 'F3', 'F4', 'F5', 'F6', 'F7', 'F8', 'F9', 'G1', 'G2', 'G3', 'G4', 'G5', 'G6', 'G7', 'G8', 'G9', 'H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'H7', 'H8', 'H9', 'I1', 'I2', 'I3', 'I4', 'I5', 'I6', 'I7', 'I8', 'I9']


In [5]:
print (unitlist)

[['A1', 'B1', 'C1', 'D1', 'E1', 'F1', 'G1', 'H1', 'I1'], ['A2', 'B2', 'C2', 'D2', 'E2', 'F2', 'G2', 'H2', 'I2'], ['A3', 'B3', 'C3', 'D3', 'E3', 'F3', 'G3', 'H3', 'I3'], ['A4', 'B4', 'C4', 'D4', 'E4', 'F4', 'G4', 'H4', 'I4'], ['A5', 'B5', 'C5', 'D5', 'E5', 'F5', 'G5', 'H5', 'I5'], ['A6', 'B6', 'C6', 'D6', 'E6', 'F6', 'G6', 'H6', 'I6'], ['A7', 'B7', 'C7', 'D7', 'E7', 'F7', 'G7', 'H7', 'I7'], ['A8', 'B8', 'C8', 'D8', 'E8', 'F8', 'G8', 'H8', 'I8'], ['A9', 'B9', 'C9', 'D9', 'E9', 'F9', 'G9', 'H9', 'I9'], ['A1', 'A2', 'A3', 'A4', 'A5', 'A6', 'A7', 'A8', 'A9'], ['B1', 'B2', 'B3', 'B4', 'B5', 'B6', 'B7', 'B8', 'B9'], ['C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9'], ['D1', 'D2', 'D3', 'D4', 'D5', 'D6', 'D7', 'D8', 'D9'], ['E1', 'E2', 'E3', 'E4', 'E5', 'E6', 'E7', 'E8', 'E9'], ['F1', 'F2', 'F3', 'F4', 'F5', 'F6', 'F7', 'F8', 'F9'], ['G1', 'G2', 'G3', 'G4', 'G5', 'G6', 'G7', 'G8', 'G9'], ['H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'H7', 'H8', 'H9'], ['I1', 'I2', 'I3', 'I4', 'I5', 'I6', 'I7', 'I8'

In [6]:
print (units)

{'I6': [['A6', 'B6', 'C6', 'D6', 'E6', 'F6', 'G6', 'H6', 'I6'], ['I1', 'I2', 'I3', 'I4', 'I5', 'I6', 'I7', 'I8', 'I9'], ['G4', 'G5', 'G6', 'H4', 'H5', 'H6', 'I4', 'I5', 'I6']], 'H9': [['A9', 'B9', 'C9', 'D9', 'E9', 'F9', 'G9', 'H9', 'I9'], ['H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'H7', 'H8', 'H9'], ['G7', 'G8', 'G9', 'H7', 'H8', 'H9', 'I7', 'I8', 'I9']], 'I2': [['A2', 'B2', 'C2', 'D2', 'E2', 'F2', 'G2', 'H2', 'I2'], ['I1', 'I2', 'I3', 'I4', 'I5', 'I6', 'I7', 'I8', 'I9'], ['G1', 'G2', 'G3', 'H1', 'H2', 'H3', 'I1', 'I2', 'I3']], 'E8': [['A8', 'B8', 'C8', 'D8', 'E8', 'F8', 'G8', 'H8', 'I8'], ['E1', 'E2', 'E3', 'E4', 'E5', 'E6', 'E7', 'E8', 'E9'], ['D7', 'D8', 'D9', 'E7', 'E8', 'E9', 'F7', 'F8', 'F9']], 'H3': [['A3', 'B3', 'C3', 'D3', 'E3', 'F3', 'G3', 'H3', 'I3'], ['H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'H7', 'H8', 'H9'], ['G1', 'G2', 'G3', 'H1', 'H2', 'H3', 'I1', 'I2', 'I3']], 'H7': [['A7', 'B7', 'C7', 'D7', 'E7', 'F7', 'G7', 'H7', 'I7'], ['H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'H7', 'H8', 'H9'], 

In [159]:
for u in units['B6']:
    print "u:{}".format(u)
    dplaces = [s for s in u]
    print "dplaces:{}".format(dplaces)


u:['A6', 'B6', 'C6', 'D6', 'E6', 'F6', 'G6', 'H6', 'I6']
dplaces:['A6', 'B6', 'C6', 'D6', 'E6', 'F6', 'G6', 'H6', 'I6']
u:['B1', 'B2', 'B3', 'B4', 'B5', 'B6', 'B7', 'B8', 'B9']
dplaces:['B1', 'B2', 'B3', 'B4', 'B5', 'B6', 'B7', 'B8', 'B9']
u:['A4', 'A5', 'A6', 'B4', 'B5', 'B6', 'C4', 'C5', 'C6']
dplaces:['A4', 'A5', 'A6', 'B4', 'B5', 'B6', 'C4', 'C5', 'C6']


In [7]:
print (peers)

{'I6': set(['I9', 'G6', 'G5', 'G4', 'F6', 'I8', 'I1', 'H6', 'I3', 'I2', 'I5', 'I4', 'I7', 'H5', 'B6', 'H4', 'A6', 'D6', 'E6', 'C6']), 'H9': set(['I9', 'G7', 'H8', 'D9', 'I8', 'H2', 'F9', 'H1', 'H6', 'H7', 'I7', 'H5', 'C9', 'E9', 'G9', 'H3', 'A9', 'G8', 'B9', 'H4']), 'I2': set(['I9', 'I8', 'F2', 'G3', 'G2', 'G1', 'H2', 'H3', 'I3', 'H1', 'I5', 'I4', 'I7', 'I6', 'I1', 'A2', 'B2', 'C2', 'D2', 'E2']), 'E8': set(['I8', 'H8', 'F7', 'F8', 'F9', 'G8', 'A8', 'C8', 'E9', 'B8', 'D8', 'D9', 'E5', 'D7', 'E4', 'E6', 'E1', 'E7', 'E3', 'E2']), 'H3': set(['H8', 'F3', 'G3', 'G2', 'G1', 'H1', 'H2', 'I3', 'I2', 'H6', 'H7', 'H4', 'H5', 'I1', 'A3', 'H9', 'B3', 'C3', 'D3', 'E3']), 'H7': set(['G7', 'I8', 'H8', 'H9', 'F7', 'H2', 'H3', 'H1', 'H6', 'I7', 'H5', 'I9', 'B7', 'G9', 'A7', 'D7', 'E7', 'G8', 'C7', 'H4']), 'I7': set(['G7', 'I8', 'H8', 'H9', 'F7', 'I1', 'I3', 'I2', 'I5', 'H7', 'G9', 'G8', 'B7', 'I9', 'A7', 'D7', 'I4', 'E7', 'I6', 'C7']), 'I4': set(['I9', 'I8', 'G5', 'G4', 'F4', 'G6', 'I1', 'I5', 'I3', 'I2

In [153]:
for s,d in grid_values(grid1).items():
    print s,d

I6 0
H9 9
I2 0
E8 0
H3 0
H7 0
I7 3
I4 0
H5 0
F9 0
G7 5
G6 9
G5 0
E1 7
G3 2
G2 0
G1 0
I1 0
C8 0
I3 5
E5 0
I5 1
C9 0
G9 0
G8 0
A1 0
A3 3
A2 0
A5 2
A4 0
A7 6
A6 0
C3 1
C2 0
C1 0
E6 0
C7 4
C6 6
C5 0
C4 8
I9 0
D8 0
I8 0
E4 0
D9 0
H8 0
F6 8
A9 0
G4 6
A8 0
E7 0
E3 0
F1 0
F2 0
F3 6
F4 7
F5 0
E2 0
F7 2
F8 0
D2 0
H1 8
H6 3
H2 0
H4 2
D3 8
B4 3
B5 0
B6 5
B7 0
E9 8
B1 9
B2 0
B3 0
D6 2
D7 9
D4 1
D5 0
B8 0
B9 1
D1 0


In [71]:
parse_grid(puzzle)

{'A1': '4',
 'A2': '1679',
 'A3': '12679',
 'A4': '139',
 'A5': '2369',
 'A6': '269',
 'A7': '8',
 'A8': '1239',
 'A9': '5',
 'B1': '26789',
 'B2': '3',
 'B3': '1256789',
 'B4': '14589',
 'B5': '24569',
 'B6': '245689',
 'B7': '12679',
 'B8': '1249',
 'B9': '124679',
 'C1': '2689',
 'C2': '15689',
 'C3': '125689',
 'C4': '7',
 'C5': '234569',
 'C6': '245689',
 'C7': '12369',
 'C8': '12349',
 'C9': '123469',
 'D1': '3789',
 'D2': '2',
 'D3': '15789',
 'D4': '3459',
 'D5': '34579',
 'D6': '4579',
 'D7': '13579',
 'D8': '6',
 'D9': '13789',
 'E1': '3679',
 'E2': '15679',
 'E3': '15679',
 'E4': '359',
 'E5': '8',
 'E6': '25679',
 'E7': '4',
 'E8': '12359',
 'E9': '12379',
 'F1': '36789',
 'F2': '4',
 'F3': '56789',
 'F4': '359',
 'F5': '1',
 'F6': '25679',
 'F7': '23579',
 'F8': '23589',
 'F9': '23789',
 'G1': '289',
 'G2': '89',
 'G3': '289',
 'G4': '6',
 'G5': '459',
 'G6': '3',
 'G7': '1259',
 'G8': '7',
 'G9': '12489',
 'H1': '5',
 'H2': '6789',
 'H3': '3',
 'H4': '2',
 'H5': '479',
 '

### Helper Functions
To show the initial sudoku grid in a pretty format.

In [3]:
def assign_only(grid):
    """Convert grid to a dict of possible values, {square: digits}, or
    return False if a contradiction is detected."""
    ## To start, every square can be any digit; then assign values from the grid.
    values = dict((s, digits) for s in squares)
    for s,d in grid_values(grid).items():
        if d in digits:
            values[s] = d
    return values

def pretty_print(grid):
    """Convert grid to a dict of possible values, {square: digits}, or
    return False if a contradiction is detected."""
    ## To start, every square can be any digit; then assign values from the grid.
    values = dict((s, '.') for s in squares)
    for s,d in grid_values(grid).items():
        if d in digits:
            values[s] = d
    return values

### Constraint Propagation
The function parse_grid calls assign(values, s, d). We could implement this as values[s] = d, but we can do more than just that. Those with experience solving Sudoku puzzles know that there are two important strategies that we can use to make progress towards filling in all the squares:
1. If a square has only one possible value, then eliminate that value from the square's peers. 
1. If a unit has only one possible place for a value, then put the value there.

In [4]:
def assign(values, s, d, mode=3, debug=False):
    """Eliminate all the other values (except d) from values[s] and propagate.
    Return values, except return False if a contradiction is detected."""
    other_values = values[s].replace(d, '')
    
    if debug:
        print "Assigning {} to square {}".format(d, s)
        display(values)
    
    # remove the other values in this square
    if all(eliminate(values, s, d2, mode, debug) for d2 in other_values):
        return values
    else:
        return False

def eliminate(values, s, d, mode=3, debug=False):
    """Eliminate d from values[s]; propagate when values or places <= 2.
    Return values, except return False if a contradiction is detected."""
    if d not in values[s]:
        return values ## Already eliminated
    values[s] = values[s].replace(d,'')
    
    if debug:
        print "Eliminating {} in square {}".format(d, s)
        display(values)
    
    if mode==1 or mode==3:
        ## (1) If a square s is reduced to one value d2, then eliminate d2 from the peers.
        if len(values[s]) == 0:
            return False ## Contradiction: removed last value
        elif len(values[s]) == 1:
            d2 = values[s]
            if not all(eliminate(values, s2, d2, mode, debug) for s2 in peers[s]):
                return False

    if mode==2 or mode==3:  
        ## (2) If a unit u is reduced to only one place for a value d, then put it there.
        for u in units[s]:
            dplaces = [s for s in u if d in values[s]]
            if len(dplaces) == 0:
                return False ## Contradiction: no place for this value
            elif len(dplaces) == 1:
                # d can only be in one place in unit; assign it there
                if not assign(values, dplaces[0], d, mode, debug):
                    return False
            
    return values

In [5]:
def parse_grid(grid, mode=3, debug=False):
    """Convert grid to a dict of possible values, {square: digits}, or
    return False if a contradiction is detected."""
    ## To start, every square can be any digit; then assign values from the grid.
    values = dict((s, digits) for s in squares)
    for s,d in grid_values(grid).items():
        if d in digits and not assign(values, s, d, mode, debug):
            return False ## (Fail if we can't assign d to square s.)
    return values

def grid_values(grid):
    "Convert grid into a dict of {square: char} with '0' or '.' for empties."
    chars = [c for c in grid if c in digits or c in '0.']
    assert len(chars) == 81
    return dict(zip(squares, chars))

def display(values):
    "Display these values as a 2-D grid."
    width = 1+max(len(values[s]) for s in squares)
    
    header = '  '+''.join([d+(' '*(width-1))+('|' if d in '36' else '') for d in digits])
    print (header)
              
    line = ('  '+'+'.join(['-'*(width*3)]*3))
    print (line)
    for r in rows:
        print r + '|'+ ''.join(values[r+c].center(width)+('|' if c in '36' else '')
                      for c in cols)
        if r in 'CF': print line
    print

In [8]:
#  first example from a list of easy puzzles from the fine Project Euler Sudoku problem and tried it:
grid1 = '003020600900305001001806400008102900700000008006708200002609500800203009005010300'

display(pretty_print(grid1))

  1 2 3 |4 5 6 |7 8 9 
  ------+------+------
A|. . 3 |. 2 . |6 . . 
B|9 . . |3 . 5 |. . 1 
C|. . 1 |8 . 6 |4 . . 
  ------+------+------
D|. . 8 |1 . 2 |9 . . 
E|7 . . |. . . |. . 8 
F|. . 6 |7 . 8 |2 . . 
  ------+------+------
G|. . 2 |6 . 9 |5 . . 
H|8 . . |2 . 3 |. . 9 
I|. . 5 |. 1 . |3 . . 



In [9]:
# showing all the possible options in each square
display(assign_only(grid1))

  1         2         3         |4         5         6         |7         8         9         
  ------------------------------+------------------------------+------------------------------
A|123456789 123456789     3     |123456789     2     123456789 |    6     123456789 123456789 
B|    9     123456789 123456789 |    3     123456789     5     |123456789 123456789     1     
C|123456789 123456789     1     |    8     123456789     6     |    4     123456789 123456789 
  ------------------------------+------------------------------+------------------------------
D|123456789 123456789     8     |    1     123456789     2     |    9     123456789 123456789 
E|    7     123456789 123456789 |123456789 123456789 123456789 |123456789 123456789     8     
F|123456789 123456789     6     |    7     123456789     8     |    2     123456789 123456789 
  ------------------------------+------------------------------+------------------------------
G|123456789 123456789     2     |    6     1234567

In [34]:
# In this case, the puzzle was completely solved by rote application of strategies (1) and (2)!
display(parse_grid(grid1,debug=True))

Assigning 9 to square H9
  1         2         3         |4         5         6         |7         8         9         
  ------------------------------+------------------------------+------------------------------
A|123456789 123456789 123456789 |123456789 123456789 123456789 |123456789 123456789 123456789 
B|123456789 123456789 123456789 |123456789 123456789 123456789 |123456789 123456789 123456789 
C|123456789 123456789 123456789 |123456789 123456789 123456789 |123456789 123456789 123456789 
  ------------------------------+------------------------------+------------------------------
D|123456789 123456789 123456789 |123456789 123456789 123456789 |123456789 123456789 123456789 
E|123456789 123456789 123456789 |123456789 123456789 123456789 |123456789 123456789 123456789 
F|123456789 123456789 123456789 |123456789 123456789 123456789 |123456789 123456789 123456789 
  ------------------------------+------------------------------+------------------------------
G|123456789 123456789 123

H| 12345678  12345678  12345678 | 12345678  12345678  12345678 |  124678    124678      9     
I| 12456789  12456789  12456789 | 1245678   1245678   1245678  |    3       124678    124678  

Eliminating 9 in square A6
  1         2         3         |4         5         6         |7         8         9         
  ------------------------------+------------------------------+------------------------------
A|123456789 123456789 123456789 |123456789 123456789  12345678 | 1246789  123456789  12345678 
B|123456789 123456789 123456789 |123456789 123456789  12345678 | 1246789  123456789  12345678 
C|123456789 123456789 123456789 |123456789 123456789 123456789 | 1246789  123456789  12345678 
  ------------------------------+------------------------------+------------------------------
D|123456789 123456789 123456789 |123456789 123456789 123456789 | 1246789  123456789  12345678 
E|123456789 123456789 123456789 |123456789 123456789  12345678 | 1246789  123456789  12345678 
F|123456789 123456789 

  1         2         3         |4         5         6         |7         8         9         
  ------------------------------+------------------------------+------------------------------
A| 12345689 123456789    389    |123456789  23456789  12345678 | 1246789  123456789  12345678 
B| 12345689 123456789  1346789  |123456789  23456789  12345678 | 1246789  123456789  12345678 
C| 12345689 123456789  1346789  |123456789  23456789  12345678 | 1246789  123456789  12345678 
  ------------------------------+------------------------------+------------------------------
D| 12345689  12345689   134689  |123456789  23456789  12345678 | 1246789  123456789  12345678 
E|    7      12345689   134689  | 12345689  2345689   1234568  |  124689   12345689  1234568  
F| 12345689  12345689   134689  |123456789  23456789  12345678 | 1246789  123456789  12345678 
  ------------------------------+------------------------------+------------------------------
G|  13468     134678      2     |  34678     34678

F| 12345689  12345689    4689   |123456789  3456789   12345678 |  12789   123456789  12345678 
  ------------------------------+------------------------------+------------------------------
G|  13468     134678      2     |  34678     34678       9     |    5       14678     14678   
H|  13468     134678     4678   | 2345678    345678   2345678  |   1278     124678      9     
I|   4689     46789       5     |  24678       1       24678   |    3       24678     24678   

Assigning 4 to square C7
  1         2         3         |4         5         6         |7         8         9         
  ------------------------------+------------------------------+------------------------------
A|   4589     45789       3     |  145789      2       14578   |    6       15789      1578   
B|  245689   2456789    46789   | 13456789  3456789   1345678  |  12789    1235789    123578  
C|  25689     256789      1     |  356789    356789    35678   |    4       235789    23578   
  ----------------------

  1         2         3         |4         5         6         |7         8         9         
  ------------------------------+------------------------------+------------------------------
A|   4589     45789       3     |  14579       2        1457   |    6       15789      1578   
B|  245689   2456789    46789   |  134579    34579     13457   |  12789    1235789    123578  
C|   259       2579       1     |    8        3579       6     |    4       23579      2357   
  ------------------------------+------------------------------+------------------------------
D| 12345689  1234589     4689   | 1234579    345679    123457  |  12789   123456789  12345678 
E|    7      12345689    4689   |  123459    34569     12345   |   1289    12345689  1234568  
F|  123459    123459      6     | 1234579    34579       8     |   1279    1234579   1234567  
  ------------------------------+------------------------------+------------------------------
G|   1348     13478       2     |    6        3478

D|  123459  1234589    489   |  12359    34569     1245  |   1789   13456789 1345678 
E|    7     1234589    489   |  12359    34569     1245  |   189    1345689   134568 
F|  13459    13459      6    |    7       3459      8    |    2      13459     1345  
  ---------------------------+---------------------------+---------------------------
G|   134      1347      2    |    6        8        9    |    5       147      147   
H|    8       1467      47   |    2        57       3    |    17      1467      9    
I|    69      679       5    |    4        1        7    |    3       268      268   

Eliminating 7 in square I2
  1        2        3        |4        5        6        |7        8        9        
  ---------------------------+---------------------------+---------------------------
A|   459     45789      3    |   159       2       1457  |    6      15789     1578  
B|  24569   2456789    4789  |   1359    34579     1457  |   1789   1235789   123578 
C|   259      2579      1 

G|  134    1347    2   |   6      8      9   |   5     147    147  
H|   8     147     47  |   2      5      3   |   17     6      9   
I|   69     69     5   |   4      1      7   |   3      8      2   

Eliminating 7 in square B7
  1      2      3      |4      5      6      |7      8      9      
  ---------------------+---------------------+---------------------
A|  459     8      3   |   19     2      14  |   6     1579   157  
B|  2469  24679   479  |   3     479     5   |   89    1279   178  
C|  259    2579    1   |   8      79     6   |   4    23579   357  
  ---------------------+---------------------+---------------------
D| 123459 123459   8   |  159    349    124  |  179   134579   6   
E|   7    123459   49  |  159     6     124  |   19   13459    8   
F| 13459  13459    6   |   7     349     8   |   2    13459   1345 
  ---------------------+---------------------+---------------------
G|  134    1347    2   |   6      8      9   |   5     147    147  
H|   8     147     4

G|   3      17     2   |   6      8      9   |   5     147     47  
H|   8      1      4   |   2      5      3   |   7      6      9   
I|   6      9      5   |   4      1      7   |   3      8      2   

Eliminating 1 in square D1
  1      2      3      |4      5      6      |7      8      9      
  ---------------------+---------------------+---------------------
A|   45     8      3   |   19     2      14  |   6     579     57  
B|   9      6      47  |   3      47     5   |   8      2      1   
C|  259    257     1   |   8      79     6   |   4     3579   357  
  ---------------------+---------------------+---------------------
D|  2459  12345    8   |  159    349    124  |  179   134579   6   
E|   7    12345    9   |   15     6     124  |   1     1345    8   
F|   1     345     6   |   7     349     8   |   2     3459   345  
  ---------------------+---------------------+---------------------
G|   3      17     2   |   6      8      9   |   5     147     47  
H|   8      1      4

  ------------+------------+------------
G| 3   7   2  | 6   8   9  | 5   1   4  
H| 8   1   4  | 2   5   3  | 7   6   9  
I| 6   9   5  | 4   1   7  | 3   8   2  

Assigning 4 to square B5
  1   2   3   |4   5   6   |7   8   9   
  ------------+------------+------------
A| 4   8   3  | 9   2   1  | 6   5   7  
B| 9   6   7  | 3   4   5  | 8   2   1  
C| 25  25  1  | 8   7   6  | 4   9   3  
  ------------+------------+------------
D| 25 245  8  | 1   3   24 | 9   7   6  
E| 7  234  9  | 5   6   24 | 1   34  8  
F| 1   34  6  | 7   9   8  | 2   34  5  
  ------------+------------+------------
G| 3   7   2  | 6   8   9  | 5   1   4  
H| 8   1   4  | 2   5   3  | 7   6   9  
I| 6   9   5  | 4   1   7  | 3   8   2  

Assigning 4 to square A1
  1   2   3   |4   5   6   |7   8   9   
  ------------+------------+------------
A| 4   8   3  | 9   2   1  | 6   5   7  
B| 9   6   7  | 3   4   5  | 8   2   1  
C| 25  25  1  | 8   7   6  | 4   9   3  
  ------------+------------+------------
D| 25

In [8]:
# Unfortunately, that will not always be the case. Here is the first example from a list of hard puzzles:
grid2 = '4.....8.5.3..........7......2.....6.....8.4......1.......6.3.7.5..2.....1.4......'

display(parse_grid(grid2))

  1       2       3       |4       5       6       |7       8       9       
  ------------------------+------------------------+------------------------
A|   4      1679   12679  |  139     2369    269   |   8      1239     5    
B| 26789     3    1256789 | 14589   24569   245689 | 12679    1249   124679 
C|  2689   15689   125689 |   7     234569  245689 | 12369   12349   123469 
  ------------------------+------------------------+------------------------
D|  3789     2     15789  |  3459   34579    4579  | 13579     6     13789  
E|  3679   15679   15679  |  359      8     25679  |   4     12359   12379  
F| 36789     4     56789  |  359      1     25679  | 23579   23589   23789  
  ------------------------+------------------------+------------------------
G|  289      89     289   |   6      459      3    |  1259     7     12489  
H|   5      6789     3    |   2      479      1    |   69     489     4689  
I|   1      6789     4    |  589     579     5789  | 23569   23589   23689  

### Search
There are two choices we have to make in implementing the search: 
1. variable ordering (which square do we try first?) and value ordering (which digit do we try first for the square?). For variable ordering, we will use a common heuristic called minimum remaining values, which means that we choose the (or one of the) square with the minimum number of possible values. 
Why? Consider grid2 above. Suppose we chose B3 first. It has 7 possibilities (1256789), so we'd expect to guess wrong with probability 6/7. If instead we chose G2, which only has 2 possibilities (89), we'd expect to be wrong with probability only 1/2. 
Thus we choose the square with the fewest possibilities and the best chance of guessing right. 

1. For value ordering we won't do anything special; we'll consider the digits in numeric order.

In [6]:
def solve(grid, debug=False): return search(parse_grid(grid, debug=debug), debug)

def search(values, debug=False):
    "Using depth-first search and propagation, try all possible values."
    if values is False:
        return False ## Failed earlier
    if all(len(values[s]) == 1 for s in squares): 
        return values ## Solved!
    ## Chose the unfilled square s with the fewest possibilities
    n,s = min((len(values[s]), s) for s in squares if len(values[s]) > 1)
    
    if debug:
        print "Choosing {} with the min possibilities of {}".format(s, n)
    
    return some(search(assign(values.copy(), s, d, debug=debug), debug) for d in values[s])

def some(seq):
    "Return some element of seq that is true."
    for e in seq:
        if e: return e
    return False

In [9]:
display(solve(grid2, debug=True))
print ("Solved in %.3f secs" % time_solve(grid2)[0])

Assigning 3 to square G6
  1         2         3         |4         5         6         |7         8         9         
  ------------------------------+------------------------------+------------------------------
A|123456789 123456789 123456789 |123456789 123456789 123456789 |123456789 123456789 123456789 
B|123456789 123456789 123456789 |123456789 123456789 123456789 |123456789 123456789 123456789 
C|123456789 123456789 123456789 |123456789 123456789 123456789 |123456789 123456789 123456789 
  ------------------------------+------------------------------+------------------------------
D|123456789 123456789 123456789 |123456789 123456789 123456789 |123456789 123456789 123456789 
E|123456789 123456789 123456789 |123456789 123456789 123456789 |123456789 123456789 123456789 
F|123456789 123456789 123456789 |123456789 123456789 123456789 |123456789 123456789 123456789 
  ------------------------------+------------------------------+------------------------------
G|123456789 123456789 123

H| 23456789 123456789 123456789 | 12456789  12456789  12456789 |123456789 123456789 123456789 
I|    1     123456789 123456789 | 12456789  12456789  12456789 |123456789 123456789  23456789 

Eliminating 1 in square I8
  1         2         3         |4         5         6         |7         8         9         
  ------------------------------+------------------------------+------------------------------
A|123456789 123456789 123456789 |123456789 123456789  12456789 |123456789 123456789 123456789 
B|123456789 123456789 123456789 |123456789 123456789  12456789 |123456789 123456789 123456789 
C|123456789 123456789 123456789 |123456789 123456789  12456789 |123456789 123456789 123456789 
  ------------------------------+------------------------------+------------------------------
D|123456789 123456789 123456789 |123456789 123456789  12456789 |123456789 123456789 123456789 
E|123456789 123456789 123456789 |123456789 123456789  12456789 |123456789 123456789 123456789 
F| 23456789 123456789 


Eliminating 4 in square E5
  1         2         3         |4         5         6         |7         8         9         
  ------------------------------+------------------------------+------------------------------
A| 23456789 123456789  12356789 |123456789 123456789  12456789 |123456789 123456789 123456789 
B| 23456789 123456789  12356789 |123456789 123456789  12456789 |123456789 123456789 123456789 
C| 23456789 123456789  12356789 |123456789 123456789  12456789 |123456789 123456789 123456789 
  ------------------------------+------------------------------+------------------------------
D| 23456789 123456789  12356789 |123456789 123456789  12456789 |123456789 123456789 123456789 
E| 23456789 123456789  12356789 |123456789   56789    12456789 |123456789 123456789 123456789 
F| 23456789 123456789  12356789 |123456789 123456789  12456789 |123456789 123456789 123456789 
  ------------------------------+------------------------------+------------------------------
G|  256789    256789  

H| 2356789   2356789   2356789  | 12456789  1245679   12456789 | 1234569   12345689  12345689 
I|    1      2356789      4     |  256789    25679     256789  |  23569     235689    235689  

Eliminating 6 in square C4
  1         2         3         |4         5         6         |7         8         9         
  ------------------------------+------------------------------+------------------------------
A|    4      1235679   1235679  | 1235679   1235679    125679  |    8       123569   1235679  
B| 2356789   12356789  12356789 |123456789  12345679  12456789 | 12345679  1234569   12345679 
C| 2356789   12356789  12356789 |   789     12345679  12456789 | 12345679  1234569   12345679 
  ------------------------------+------------------------------+------------------------------
D| 2356789  123456789  12356789 | 12345679  12345679  1245679  | 12345679  12345689 123456789 
E|  235679   12345679  1235679  | 12345679     8      1245679  | 12345679  1234569   12345679 
F| 2356789  123456789 

  ------------------------------+------------------------------+------------------------------
D|  235789   12345789  1235789  |  123459   1234579    124579  |  123579      6      1234789  
E|  235679   12345679  1235679  |  123459      8      1245679  |    4       123459    123479  
F| 2356789  123456789  12356789 |  123459   12345679  1245679  |  123579    123589    123789  
  ------------------------------+------------------------------+------------------------------
G|   2589      2589      2589   |    6       12459       3     |   1259       7       12489   
H| 2356789   2356789   2356789  |  124589    124579   1245789  |  123569   1234589   1234689  
I|    1      2356789      4     |   2589      2579     25789   |  23569     23589     23689   

Assigning 4 to square E7
  1         2         3         |4         5         6         |7         8         9         
  ------------------------------+------------------------------+------------------------------
A|    4       123679    

  ---------------------------+---------------------------+---------------------------
G|   289       89      289   |    6       2459      3    |   1259      7      12489  
H|    5      36789    236789 |  12489     2479    124789 |  12369    123489  1234689 
I|    1      36789      4    |   2589     2579    25789  |  23569    23589    23689  

Eliminating 1 in square H4
  1        2        3        |4        5        6        |7        8        9        
  ---------------------------+---------------------------+---------------------------
A|    4      13679    123679 |   1239     2369     1269  |    8       1239      5    
B|  236789  1356789  12356789| 1234589   234569  1245689 |  123679   12349   1234679 
C|  23689    135689  1235689 |    7      234569  1245689 |  12369    12349    123469 
  ---------------------------+---------------------------+---------------------------
D|   3789      2      135789 |   3459    34579     4579  |  13579      6      13789  
E|   3679    135679   1356

G|   2      8      9   |   6      4      3   |   5      7      1   
H|   5      67     3   |   2     479     1   |   69    489    4689 
I|   1      67     4   |  589    579    5789 |  2369   2389  23689 

Assigning 1 to square G9
  1      2      3      |4      5      6      |7      8      9      
  ---------------------+---------------------+---------------------
A|   4     1679   1267 |  139    2369   269  |   8     1239    5   
B|  6789    3    125678| 14589  24569  245689| 12679   1249  24679 
C|  689    1569  12568 |   7    234569 245689| 12369  12349  23469 
  ---------------------+---------------------+---------------------
D|  3789    2     1578 |  3459  34579   4579 |  1379    6     3789 
E|  3679  15679   1567 |  359     8    25679 |   4    12359   2379 
F| 36789    4     5678 |  359     1    25679 |  2379  23589  23789 
  ---------------------+---------------------+---------------------
G|   2      8      9   |   6      4      3   |   5      7      1   
H|   5      67     3  

  ---------------------+---------------------+---------------------
G|   2      8      9   |   6      4      3   |   5      7      1   
H|   5      6      3   |   2      7      1   |   9      48     48  
I|   1      7      4   |   8      59     59  |  236     23    236  

Eliminating 2 in square I8
  1      2      3      |4      5      6      |7      8      9      
  ---------------------+---------------------+---------------------
A|   4      1      7   |   3     269    269  |   8      2      5   
B|  689     3     2568 |   1     2569  245689|  267    249   24679 
C|  689     59    2568 |   7     2569  245689|  1236  12349  23469 
  ---------------------+---------------------+---------------------
D|  789     2     158  |   4      3     579  |   17     6     789  
E|  3679    59    156  |   59     8    25679 |   4    12359   2379 
F| 36789    4     568  |   59     1    25679 |  237   23589  23789 
  ---------------------+---------------------+---------------------
G|   2      8      9

H|   5      6      3   |   2      7      1   |   9      48     48  
I|   1      7      4   |   8      59     59  |   26     3      26  

Eliminating 6 in square A5
  1      2      3      |4      5      6      |7      8      9      
  ---------------------+---------------------+---------------------
A|   4      1      7   |   3      9      69  |   8      2      5   
B|  689     3     2568 |   1     2569  245689|   67     49    4679 
C|  689     59    2568 |   7     2569  245689|  136    149    3469 
  ---------------------+---------------------+---------------------
D|  789     2     158  |   4      3     579  |   17     6     789  
E|  3679    59    156  |   59     8    25679 |   4     159    2379 
F| 36789    4     568  |   59     1    25679 |  237    589   23789 
  ---------------------+---------------------+---------------------
G|   2      8      9   |   6      4      3   |   5      7      1   
H|   5      6      3   |   2      7      1   |   9      48     48  
I|   1      7      4

A|  4     1     7   |  3     2     6   |  8     9     5   
B| 689    3    268  |  1    569  45689 | 267    24   2467 
C| 689    5    268  |  7    2569  4689 | 236    1    2346 
  ------------------+------------------+------------------
D| 789    2     5   |  4     3     79  |  1     6    789  
E| 3679   9     1   |  59    8     69  |  4    235   237  
F| 3678   4     68  |  59    1     2   |  37   358   3789 
  ------------------+------------------+------------------
G|  2     8     9   |  6     4     3   |  5     7     1   
H|  5     6     3   |  2     7     1   |  9     48    48  
I|  1     7     4   |  8     59    59  | 236    23   236  

Eliminating 9 in square E4
  1     2     3     |4     5     6     |7     8     9     
  ------------------+------------------+------------------
A|  4     1     7   |  3     2     6   |  8     9     5   
B| 689    3    268  |  1    569  45689 | 267    24   2467 
C| 689    5    268  |  7    2569  4689 | 236    1    2346 
  ------------------+-------

  ------------------+------------------+------------------
D| 3789   2    158  |  4    359   579  | 137    6    3789 
E| 3679   59   156  |  35    8    279  |  4   12359  2379 
F|36789   4     58  |  35    1     6   | 237  23589 23789 
  ------------------+------------------+------------------
G|  2     8     9   |  6     4     3   |  5     7     1   
H|  5     6     3   |  2     7     1   |  9     48    48  
I|  1     7     4   |  8     59    59  | 236    23   236  

Eliminating 6 in square F1
  1     2     3     |4     5     6     |7     8     9     
  ------------------+------------------+------------------
A|  4     1     7   |  9    236    2   |  8     23    5   
B| 689    3    2568 |  1    256  24568 | 267   249  24679 
C| 689    59   2568 |  7    2356 24568 | 1236 12349 23469 
  ------------------+------------------+------------------
D| 3789   2    158  |  4    359   579  | 137    6    3789 
E| 3679   59   156  |  35    8    279  |  4   12359  2379 
F| 3789   4     58  |  35   

  ------------------+------------------+------------------
G|  2     8     9   |  6     4     3   |  5     7     1   
H|  5     6     3   |  2     7     1   |  9     48    48  
I|  1     7     4   | 589    59   589  | 236   238   2368 

Eliminating 9 in square D6
  1     2     3     |4     5     6     |7     8     9     
  ------------------+------------------+------------------
A|  4     9     7   |  1    236    2   |  8     23    5   
B|  68    3   12568 | 589   2569 24589 | 1267  1249 24679 
C|  68    15  12568 |  7   23569 24589 | 1236 12349 23469 
  ------------------+------------------+------------------
D| 3789   2    158  |  4    359    7   | 137    6    3789 
E| 3679   15   156  | 359    8     29  |  4   12359  2379 
F| 3789   4     58  | 359    1     6   | 237  23589 23789 
  ------------------+------------------+------------------
G|  2     8     9   |  6     4     3   |  5     7     1   
H|  5     6     3   |  2     7     1   |  9     48    48  
I|  1     7     4   | 589   

B|   68     3     2568 |   1     2569  245689|  267    249   24679 
C|   68     15   12568 |   7     2569  245689|  1236  12349  23469 
  ---------------------+---------------------+---------------------
D|  789     2     158  |   4      3     579  |   17     6     789  
E|  3679    15    156  |   59     8    25679 |   4    12359   2379 
F| 36789    4     568  |   59     1    25679 |  237   23589  23789 
  ---------------------+---------------------+---------------------
G|   2      8      9   |   6      4      3   |   5      7      1   
H|   5      6      3   |   2      7      1   |   9      48     48  
I|   1      7      4   |   8      59     59  |  236     23    236  

Eliminating 1 in square C8
  1      2      3      |4      5      6      |7      8      9      
  ---------------------+---------------------+---------------------
A|   4      9      7   |   3      26     26  |   8      1      5   
B|   68     3     2568 |   1     2569  245689|  267    249   24679 
C|   68     15   125


Assigning 8 to square D9
  1     2     3     |4     5     6     |7     8     9     
  ------------------+------------------+------------------
A|  4     9     7   |  3     6     2   |  8     1     5   
B|  68    3    256  |  1    259  24589 | 267   249   2679 
C|  68    15   1256 |  7    2569 24589 | 236   2349  2369 
  ------------------+------------------+------------------
D|  79    2     5   |  4     3     79  |  1     6     8   
E| 3679   15   156  |  59    8     29  |  4    2359  2379 
F| 379    4     8   |  59    1     6   | 237   2359  2379 
  ------------------+------------------+------------------
G|  2     8     9   |  6     4     3   |  5     7     1   
H|  5     6     3   |  2     7     1   |  9     8     4   
I|  1     7     4   |  8     59    59  | 236    23   236  

Eliminating 5 in square B3
  1     2     3     |4     5     6     |7     8     9     
  ------------------+------------------+------------------
A|  4     9     7   |  3     6     2   |  8     1     5   
B|

Eliminating 8 in square B4
  1      2      3      |4      5      6      |7      8      9      
  ---------------------+---------------------+---------------------
A|   4      1      7   |   39    236    269  |   8     239     5   
B|  689     3     2568 |   19    256   245689|  1279   1249  24679 
C|  689     59    2568 |   7     2356  245689|  1239  12349  23469 
  ---------------------+---------------------+---------------------
D|  3789    2     158  |   4     357    579  |  1379    6     3789 
E|  3679    59    156  |  359     8    25679 |   4    12359   2379 
F| 36789    4     568  |  359     1    25679 |  2379  23589  23789 
  ---------------------+---------------------+---------------------
G|   2      8      9   |   6      4      3   |   5      7      1   
H|   5      7      3   |   2      9      1   |   6      48     48  
I|   1      6      4   |   58     57    578  |  239    2389   2389 

Assigning 8 to square I4
  1      2      3      |4      5      6      |7      8      9  

G|  2    8    9  |  6    4    3  |  5    7    1  
H|  5    7    3  |  2    9    1  |  6    48   48 
I|  1    6    4  |  8    7    5  | 239  239   23 

Eliminating 6 in square A6
  1    2    3    |4    5    6    |7    8    9    
  ---------------+---------------+---------------
A|  4    1    7  |  3    26   9  |  8   239   5  
B| 689   3   268 |  1    5   489 | 279  249  2467
C| 689   5   268 |  7    26  4689| 239   1   2346
  ---------------+---------------+---------------
D|  8    2    5  |  4    3    7  |  1    6    9  
E| 3679  9    1  |  5    8    6  |  4   235  237 
F|  37   4    6  |  9    1    2  |  37  358  378 
  ---------------+---------------+---------------
G|  2    8    9  |  6    4    3  |  5    7    1  
H|  5    7    3  |  2    9    1  |  6    48   48 
I|  1    6    4  |  8    7    5  | 239  239   23 

Eliminating 9 in square B6
  1    2    3    |4    5    6    |7    8    9    
  ---------------+---------------+---------------
A|  4    1    7  |  3    26   9  |  8   239 

Assigning 6 to square C9
  1 2 3 |4 5 6 |7 8 9 
  ------+------+------
A|4 1 7 |3 6 9 |8 2 5 
B|6 3 2 |1 5 8 |9 4 7 
C|9 5 8 |7 2 4 |3 1 6 
  ------+------+------
D|8 2 5 |4 3 7 |1 6 9 
E|7 9 1 |5 8 6 |4 3 2 
F|3 4 6 |9 1 2 |7 5 8 
  ------+------+------
G|2 8 9 |6 4 3 |5 7 1 
H|5 7 3 |2 9 1 |6 8 4 
I|1 6 4 |8 7 5 |2 9 3 

Assigning 6 to square B1
  1 2 3 |4 5 6 |7 8 9 
  ------+------+------
A|4 1 7 |3 6 9 |8 2 5 
B|6 3 2 |1 5 8 |9 4 7 
C|9 5 8 |7 2 4 |3 1 6 
  ------+------+------
D|8 2 5 |4 3 7 |1 6 9 
E|7 9 1 |5 8 6 |4 3 2 
F|3 4 6 |9 1 2 |7 5 8 
  ------+------+------
G|2 8 9 |6 4 3 |5 7 1 
H|5 7 3 |2 9 1 |6 8 4 
I|1 6 4 |8 7 5 |2 9 3 

Assigning 9 to square B7
  1 2 3 |4 5 6 |7 8 9 
  ------+------+------
A|4 1 7 |3 6 9 |8 2 5 
B|6 3 2 |1 5 8 |9 4 7 
C|9 5 8 |7 2 4 |3 1 6 
  ------+------+------
D|8 2 5 |4 3 7 |1 6 9 
E|7 9 1 |5 8 6 |4 3 2 
F|3 4 6 |9 1 2 |7 5 8 
  ------+------+------
G|2 8 9 |6 4 3 |5 7 1 
H|5 7 3 |2 9 1 |6 8 4 
I|1 6 4 |8 7 5 |2 9 3 

Assigning 7 to square F7


NameError: name 'time_solve' is not defined

In [27]:
display(solve(grid1))
print ("Solved in %.3f secs" % time_solve(grid1)[0])

  1 2 3 |4 5 6 |7 8 9 
  ------+------+------
A|4 8 3 |9 2 1 |6 5 7 
B|9 6 7 |3 4 5 |8 2 1 
C|2 5 1 |8 7 6 |4 9 3 
  ------+------+------
D|5 4 8 |1 3 2 |9 7 6 
E|7 2 9 |5 6 4 |1 3 8 
F|1 3 6 |7 9 8 |2 4 5 
  ------+------+------
G|3 7 2 |6 8 9 |5 1 4 
H|8 1 4 |2 5 3 |7 6 9 
I|6 9 5 |4 1 7 |3 8 2 

Solved in 0.014 secs


In [29]:
import time

def solve_all(grids, name=''):
    """Attempt to solve a sequence of grids. Report results."""
    times, results = zip(*[time_solve(grid) for grid in grids])
    N = len(results)
    if N > 1:
        print("Solved %d of %d %s puzzles (avg %.3f secs (%d Hz), max %.3f secs)." % 
              (sum(results), N, name, sum(times)/N, N/sum(times), max(times)))
            
def time_solve(grid):
    start = time.clock()
    values = solve(grid)
    t = time.clock()-start
    return (t, solved(values))

def solved(values):
    "A puzzle is solved if each unit is a permutation of the digits 1 to 9."
    def unitsolved(unit): return set(values[s] for s in unit) == set(digits)
    return values is not False and all(unitsolved(unit) for unit in unitlist)

In [30]:
grid1  = '003020600900305001001806400008102900700000008006708200002609500800203009005010300'
grid2  = '4.....8.5.3..........7......2.....6.....8.4......1.......6.3.7.5..2.....1.4......'
hard1  = '.....6....59.....82....8....45........3........6..3.54...325..6..................'
    
if __name__ == '__main__':
    test()
    solve_all(open("sudoku-easy50.txt"), "easy")
    solve_all(open("sudoku-top95.txt"), "hard")
    solve_all(open("sudoku-hardest.txt"), "hardest")

All tests pass.
Solved 50 of 50 easy puzzles (avg 0.004 secs (231 Hz), max 0.007 secs).
Solved 95 of 95 hard puzzles (avg 0.017 secs (59 Hz), max 0.096 secs).
Solved 11 of 11 hardest puzzles (avg 0.006 secs (155 Hz), max 0.009 secs).
